src/Microsoft.ML.CpuMath/FactorizationMachine/AvxIntrinsics.cs

```diff
@@ -0,0 +1,188 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics;
+using System.Runtime.Intrinsics.X86;
+
+namespace Microsoft.ML.Internal.CpuMath.FactorizationMachine
+{
+    internal static class AvxIntrinsics
+    {
+        // This function implements Algorithm 1 in https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf.
+        // Compute the output value of the field-aware factorization, as the sum of the linear part and the latent part.
+        // The linear part is the inner product of linearWeights and featureValues.
+        // The latent part is the sum of all intra-field interactions in one field f, for all fields possible
+        public static unsafe void CalculateIntermediateVariables(int fieldCount, int latentDim, int count, int* fieldIndices, int* featureIndices, float* featureValues,
```

Comment:
Is there any perf concern around the number of parameters being passed here?

Reply:
Can you elaborate more on this? For now, I have kept the parameters identical to the C++ code.

Reply:
C# isn't always as good about inlining as C/C+and there are more parameters being passed than will stay in registers (even for `fastcall`). Was wondering if that was a concern since these most of these will end up shadow copied to the stack and be double-indirections.

Reply:
if it is, and if inlining isn't a valid option, reordering the parameters so the pointers are first would ensure they are the ones passed in register (keeping them single indirections).

Reply:
There is a very slight difference in the codegen. Changing the order of parameters reduces one instruction. I can reorder the parameters in the code.